lib/charms/tempo_k8s/v1/charm_tracing.py

```diff
@@ -172,14 +172,62 @@ def my_tracing_endpoint(self) -> Optional[str]:
 provide an *absolute* path to the certificate file instead.
 """
 
+
+def _remove_stale_otel_sdk_packages():
+    """Hack to remove stale opentelemetry sdk packages from the charm's python venv.
+
+    See https://github.com/canonical/grafana-agent-operator/issues/146 and
+    https://bugs.launchpad.net/juju/+bug/2058335 for more context. This patch can be removed after
+    this juju issue is resolved and sufficient time has passed to expect most users of this library
+    have migrated to the patched version of juju.  When this patch is removed, un-ignore rule E402 for this file in the pyproject.toml (see setting
+    [tool.ruff.lint.per-file-ignores] in pyproject.toml).
+
+    This only has an effect if executed on an upgrade-charm event.
+    """
+    import os
+
+    if os.getenv("JUJU_DISPATCH_PATH") == "hooks/upgrade-charm":
+        # Imports done within this function so it is standalone, minimal side-effects, and easy to revert later
+        import logging
+        import shutil
+
+        from importlib_metadata import distributions
+
+        otel_logger = logging.getLogger("tracing")
+        otel_logger.debug("Executing _remove_stale_otel_sdk_packages patch on charm upgrade")
+        # Find any opentelemetry_* distributions
+        otel_distributions = {}
+        for d in distributions():
+            if d._normalized_name.startswith("opentelemetry_"):
```

Comment:
I'm surprised the private name access doesn't give a linter/pyright error. 
I'd rather do that once:

```
otel_distributions = defaultdict(list)
for distribution in distributions():   # no one-letter variables
    normalized_name = distribution._normalized_name
    if normalized_name.startswith(...): ...
```

Reply:
agreed!  adopted your offline suggestion